### PR TITLE
analytics: fix mac not sending reports

### PR DESCRIPTION
**Pull request by dberenbaum**
Not sure if this is the best approach to fix the issue, but currently we [do not seem to be sending analytics reports for mac since March](https://iterativeai.slack.com/archives/CB41NAL8H/p1697464559356029?thread_ts=1697200360.717249&cid=CB41NAL8H). I have tested and found that it seems to be some combination of forking and the `_popen` method used. Analytics get sent successfully if I either:
1. Drop all the forking logic, OR
2. Call `main` like it does for linux.

Here I dropped all the forking and conditions here and just use `_popen` directly. This successfully sends reports for me, and I'm not sure why we need forking here since it should happen in a separate, non-blocking process when using `_popen` (and dropping `.communicate()`). 

Not sure if I miss some context for why forking or other logic is needed (I took a look back at https://github.com/iterative/dvc/issues/4294 and https://github.com/iterative/dvc/pull/4262 but wasn't clear why this is all needed).

### CODE REVIEW

File: dvc/daemon.py [hunk 1]
```diff
@@ -55,62 +54,13 @@ def _spawn_windows(cmd, env):
...
-        main(cmd)
-
-    os._exit(0)  # pylint: disable=protected-access
-
-
 def _spawn(cmd, env):
     logger.debug("Trying to spawn '%s'", cmd)
 
     if os.name == "nt":
         _spawn_windows(cmd, env)
```
Comment (dberenbaum) [inline dvc/daemon.py:61]: We have a check here for `sys.platform == "win32"`. Does this mean we don't report on 64-bit windows?

**Comment by skshetry** [comment]
> Not sure if I miss some context for why forking or other logic is needed

Forking is cheaper than `popen`. Let's limit the change to `macOS` for now.

**Comment by skshetry** [comment]
Also, could you please try commenting out this line and see if that is the problem? I don't remember any other recent changes than that addition.

https://github.com/iterative/dvc/blob/8ea9d3edef146c239086a7aec3cec9ead1bd32a9/dvc/daemon.py#L94

**Comment by dberenbaum** [comment]
> Also, could you please try commenting out this line and see if that is the problem? I don't remember any other recent changes than that addition.

That doesn't help. I'm wondering if it's something in a mac os release rather than on our side causing the issue.

**Comment by dberenbaum** [comment]
> Forking is cheaper than `popen`. Let's limit the change to `macOS` for now.

Done

**Comment by efiop** [comment]
> Analytics get sent successfully if I either:
>
> Drop all the forking logic, OR
> Call main like it does for linux.

So just reverting changes from https://github.com/iterative/dvc/issues/4294 (aka do the same thing we do on linux) fixes it for you, right?

**Comment by efiop** [comment]
> currently we [do not seem to be sending analytics reports for mac since March](https://iterativeai.slack.com/archives/CB41NAL8H/p1697464559356029?thread_ts=1697200360.717249&cid=CB41NAL8H)

Were you able to find any hints on why since march specifically?

**Comment by dberenbaum** [comment]
> So just reverting changes from #4294 (aka do the same thing we do on linux) fixes it for you, right?

Yes. Not sure if #4294 was related to certain versions of the mac os.

> Were you able to find any hints on why since march specifically?

No. My best guess is some mac os update.

File: dvc/daemon.py [hunk 2]
```diff
@@ -58,50 +58,50 @@ def _spawn_windows(cmd, env):
...
-    sys.stdin.close()
-    sys.stdout.close()
-    sys.stderr.close()
-    os.closerange(0, 3)
-
     if platform.system() == "Darwin":
         # workaround for MacOS bug
         # https://github.com/iterative/dvc/issues/4294
-        _popen(cmd, env=env).communicate()
+        _popen(cmd, env=env)
```
[the remark this answers is not in the document]
Comment (dberenbaum) [reply dvc/daemon.py]: Updated to drop the mac workaround

**Comment by efiop** [comment]
@dberenbaum So just to confirm I understand correctly, in this current form, the fix works for you, right? If so, this is super clean 🔥 

> Not sure if I miss some context for why forking or other logic is needed (I took a look back at https://github.com/iterative/dvc/issues/4294 and https://github.com/iterative/dvc/pull/4262 but wasn't clear why this is all needed).

This is mostly to do the analytics sending in a separate detached process, so the main one doesn't have to wait for that if the network is slow. There was a big emphasis on quick command execution in the past that now definitely feels redundant and even misguided. But since we have the mechanism already, I guess it is nice to have one...

**Comment by dberenbaum** [comment]
> So just to confirm I understand correctly, in this current form, the fix works for you, right? If so, this is super clean 🔥

Yes.

> This is mostly to do the analytics sending in a separate detached process, so the main one doesn't have to wait for that if the network is slow. There was a big emphasis on quick command execution in the past that now definitely feels redundant and even misguided. But since we have the mechanism already, I guess it is nice to have one...

My point was that popen will still be non-blocking (without `.communicate()` and running in a separate subprocess, so I don't think it should have any performance impact besides the minor one @skshetry noted that forking is cheaper than a new subprocess. Not sure it's worth all the extra code but I understand the hesitance to change more than necessary here.

**Comment by efiop** [comment]
> My point was that popen will still be non-blocking (without .communicate()) and running in a separate subprocess, so I don't think it should have any performance impact besides the minor one @skshetry noted that forking is cheaper than a new subprocess. Not sure it's worth all the extra code but I understand the hesitance to change more than necessary here.

Ah, I see. That extra code is mostly about daemonizing the process to properly detach from parent/session/tty/etc so that there are no orphans and the child is not killed when the main process finishes and stuff like that.

**Review by efiop — APPROVED** [review]
Outstanding research, thanks for fixing analytics! 🔥

**Comment by skshetry** [comment]
I now understand why it was failing. 

I added following code in #9204:
 
https://github.com/iterative/dvc/blame/8ea9d3edef146c239086a7aec3cec9ead1bd32a9/dvc/daemon.py#L94


which closes the streams in the child. So, the `popen` does not inherit any streams, and all `sys.{stdin,stdout,stderr}` are set to `None`.

Couple that with the changes to #8993, where I am not checking if the `sys.{stdout,stderr}` is not None, the daemon is failing silently:

```python
Traceback (most recent call last):
  File "/Users/saugat/Projects/dvc/dvc/__main__.py", line 15, in <module>
    from dvc.cli import main
  File "/Users/saugat/Projects/dvc/dvc/__init__.py", line 10, in <module>
    dvc.logger.setup()
  File "/Users/saugat/Projects/dvc/dvc/logger.py", line 187, in setup
    formatter = ColorFormatter(log_colors=log_colors and sys.stdout.isatty())
                                                         ^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'isatty'
```

```python
Traceback (most recent call last):
  File "/Users/saugat/Projects/dvc/dvc/__main__.py", line 26, in <module>
    ex = main(sys.argv[1:])
         ^^^^^^^^^^^^^^^^^^
  File "/Users/saugat/Projects/dvc/dvc/cli/__init__.py", line 168, in main
    if sys.stderr.closed:  # pylint: disable=using-constant-test
       ^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'closed'
```


I think we should create a pidfile, and pass it to `dvc daemon` which gets deleted if it exits cleanly.

**Comment by dberenbaum** [comment]
If it helps, here are the steps I'm using to debug:

1. Download the latest data (I have been downloading the latest hour to debug) from s3://dvc-analytics/events
2. Run the following:

```python
import dask.bag as db
import json


def flatten_json(record: dict) -> dict:
    for k, v in record["body"].items():
        if isinstance(v, dict):
            for subk, subv in v.items():
                record[f"{k}.{subk}"] = subv
        else:
            record[k] = v
    del record["body"]
    return record

def to_dict(record: tuple) -> dict:
    # Separate record into json data and path.
    jsondata, fpath = record
    # Convert to dicts.
    dictdata = json.loads(jsondata)
    dictdata["path"] = fpath.split("sample")[-1]
    # Merge dicts.
    return dictdata

b = db.read_text(<path_to_downloaded_data>, include_path=True)
b_dict = b.map(to_dict)
b_flat = b_dict.map(flatten_json)
df = b_flat.to_dataframe().compute()
df[df["user_id"] == <my_user_id>]
```

**Comment by dberenbaum** [comment]
> I added following code in #9204:
> 
> https://github.com/iterative/dvc/blob/8ea9d3edef146c239086a7aec3cec9ead1bd32a9/dvc/daemon.py#L94

@skshetry Do you recommend we drop it? I worry what other effects it could have.

**Comment by skshetry** [comment]
I am getting crashes after pulling this changes. Found some related issues in https://github.com/python/cpython/issues/105912, and in [bpo-31818](https://bugs.python.org/issue31818) unless I set `export OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES`.

> IMHO the safest fix is to not run any Python program after fork(). Using use subprocess to use fork() immmediately followed by exec(). It's not safe to execute code after fork(), many functions are not "fork safe". But I know that many applications don't care, since there is also a lot of functions which are fork safe...

> Many Mac OS X frameworks do not work reliably if you call fork but do not call exec.

So it seems like we should really be using `popen` in macOS.

**Comment by shcheklein** [comment]
@skshetry could this bug affected other platforms somehow after all? I see that the fix after all is not Darwin specific, right?

**Comment by skshetry** [comment]
> @skshetry could this bug affected other platforms somehow after all? I see that the fix after all is not Darwin specific, right?

See https://github.com/iterative/dvc/pull/10026#issuecomment-1766646157.

On macOS, we were sending analytics via a subprocess. And, in that case, `stderr`/`stdout`/`stdin` would set to `None`, as we were closing the file descriptors.

On Linux, we were running it in a (double) forked process, so the file descriptors' objects did exist, even if they were closed. Which we could handle already.

Windows wasn't affected as it has a separate implementation.

**Comment by shcheklein** [comment]
@skshetry thanks! we see a dip in analytics the same time in other platforms ... not as drastic, but still .. do you know if were doing other changes to the telemetry? overall, what is the best way to see / map all the changes we've done to the telemetry / configs that could have affected this?

**Comment by skshetry** [comment]
I don't think so. The only major change was https://github.com/iterative/dvc/commit/db70d1cc5c34516cae04460da5433a94d6371d53. But that probably should have increased telemetry.

Could that be 3.0 related, as it was a major version?

**Comment by skshetry** [comment]
FYI, today we have a test server to see whether or not analytics (and, updater) is working or not.

```console
python -m tests.func.test_daemon 8001 &
export DVC_ANALYTICS_ENDPOINT=http://127.0.0.1:8000
export DVC_UPDATER_ENDPOINT=http://127.0.0.1:8000

dvc version
```

Which will log what was received. POST request is for `analytics` and `GET` request is for updater.

**Pull request merged**